stdlib/2/__builtin__.pyi

```diff
@@ -1346,15 +1347,55 @@ def next(__i: Iterator[_T], default: _VT) -> Union[_T, _VT]: ...
 def oct(__i: Union[int, _SupportsIndex]) -> str: ...
 
 if sys.version_info >= (3, 6):
-    def open(file: Union[str, bytes, int, _PathLike[Any]], mode: str = ..., buffering: int = ..., encoding: Optional[str] = ...,
-             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
-             opener: Optional[Callable[[str, int], int]] = ...) -> IO[Any]: ...
+    # Changed in version 3.6: Support added to accept objects implementing os.PathLike.
+    _OpenFile = Union[str, bytes, int, _PathLike[Any]]
 elif sys.version_info >= (3,):
-    def open(file: Union[str, bytes, int], mode: str = ..., buffering: int = ..., encoding: Optional[str] = ...,
-             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
-             opener: Optional[Callable[[str, int], int]] = ...) -> IO[Any]: ...
+    _OpenFile = Union[str, bytes, int]
 else:
-    def open(name: Union[unicode, int], mode: unicode = ..., buffering: int = ...) -> BinaryIO: ...
+    _OpenFile = Union[unicode, int]
+
+if sys.version_info >= (3, 3):
+    @overload
+    def open(
+        file: _OpenFile,
+        mode: _OpenTextMode = ...,
+        buffering: int = ...,
+        encoding: Optional[str] = ...,
+        errors: Optional[str] = ...,
+        newline: Optional[str] = ...,
+        closefd: bool = ...,
+        opener: Optional[Callable[[str, int], int]] = ...,
+    ) -> TextIO: ...
+    @overload
+    def open(
+        file: _OpenFile,
+        mode: _OpenBinaryMode,
+        buffering: int = ...,
+        encoding: None = ...,
+        errors: None = ...,
+        newline: None = ...,
+        closefd: bool = ...,
+        opener: Optional[Callable[[str, int], int]] = ...,
+    ) -> BinaryIO: ...
+    @overload
+    def open(
+        file: _OpenFile,
+        mode: str,
+        buffering: int = ...,
+        encoding: Optional[str] = ...,
+        errors: Optional[str] = ...,
+        newline: Optional[str] = ...,
+        closefd: bool = ...,
+        opener: Optional[Callable[[str, int], int]] = ...,
+    ) -> IO[Any]: ...
+
+else:
+    @overload
+    def open(file: _OpenFile, mode: _OpenTextMode = ..., buffering: int = ...) -> TextIO: ...
+    @overload
+    def open(file: _OpenFile, mode: _OpenBinaryMode, buffering: int = ...) -> BinaryIO: ...
+    @overload
+    def open(file: _OpenFile, mode: unicode, buffering: int = ...) -> IO[Any]: ...
```

Comment:
This doesn't seem right for Python 2. I think that this should always return `BinaryIO` in Python 2, as it used to do.

Reply:
Interesting, I didn't realize the return type was the same in text mode and binary mode for Python 2. Restored the non-overloaded version.